api/v1alpha1/etcdpeer_types.go

```diff
@@ -53,6 +54,13 @@ type EtcdPeerSpec struct {
 	// instructions if it already knows where it's peers are.
 	// +optional
 	Bootstrap *Bootstrap `json:"bootstrap,omitempty"`
+
+	// VolumeClaimTemplates is a claim that pods are allowed to reference.
+	// The EtcdPeer controller will create a new PersistentVolumeClaim using the StorageClass
+	// and the Storage Resource Request in this template.
+	// That PVC will then be mounted in the Pod for this EtcdPeer
+	// and the Etcd process when it starts will persist its data to the PV bound to that PVC.
+	VolumeClaimTemplate *corev1.PersistentVolumeClaim `json:"volumeClaimTemplate,omitempty"`
```

Comment:
do you think its worth wrapping this in another struct, e.g.
```
  ...
  
  Storage *EtcdPeerStorage `json:"storage,omitempty"`
}

type EtcdPeerStorage {
  PersistentVolume *corev1.PersistentVolumeClaim `json:"volumeClaimTemplate,omitempty"`

  // space for others, filesys mounted, etc
}
```

Reply:
I've gone off the idea of re-using `corev1.PersistentVolumeClaim`.
It would be sufficient, for now to have:

```
type EtcdPeerStorage {
  StorageClassName string
  Size resource.Quantity

  // space for others, filesys mounted, etc
}
```

And copy values from there into the PVC we create.

Reply:
do you mind elaborating a bit on that? from my point of view, exposing the full pvc would allow more flexible and less opinionated configuration of the peer. for example if we wanted to point the peer to a volume created from a snapshot as part of the recovery process, that would be trivial. downside is that we'd still have to construct this pvc in the cluster controller (because i don't think we'd trivially be able to expose a pvc "template" through the cluster resource), and that would likely include some duplication of names, etc. 

is there anything i'm missing?

Reply:
> For example if we wanted to point the peer to a volume created from a snapshot as part of the recovery process, that would be trivial.

The indented logic for the peer controller, regarding PVCs, is:

> If a correctly named persistent volume claim (PVC) does not exist,
   create one with the settings provided on the peer resource.
>
> We never delete a persistent volume claim automatically, so this
   behaviour ensures that the operator will resume stopped clusters if
   the storage remains even if the peer resource was deleted. In
   future, automatic cluster recovery tools **may** delete PVC
   resources if they determine it is required and safe to do so,
   although this is outside of the scope of this document.
>
>  If the storage options on the peer resource are different from an
   existing PVC, the operator makes no attempt to 'correct' the
   preexisting PVC.

(See `docs/design/resources.md` on `master`)

So we only use this template information to create a PVC if we don't already find one with the correct name. So if you want to restore from a backup of the peer's data directory you can just add the PVC before you start the peer.

Reply:
i was using the snapshot restore as an example of an extra feature that is offered in the PVC spec, which we'd lose if we masked the fields from the peer resource:  (https://kubernetes.io/docs/concepts/storage/persistent-volumes/#volume-cloning)

its something we probably won't need for now but its conceivable that someone would want it in the future. this also applies to things like the volume mode & access modes. down the road if we ended up implementing all these options we'd just end up with a clone of `corev1.PersistentVolumeClaim`. i'd like to understand the drawbacks are of just embedding this object :)

Reply:
> i'd like to understand the drawbacks are of just embedding this object

I think we *should* be very opinionated about how the cluster gets deployed (at least for now).
If we give users the ability all the fields of `PersistentVolumeClaimSpec` then we have to add validation to prevent them setting the fields that we are not yet handling.

If we *do* allow them to configure all those fields, then we have to be able to test  and support Cluster deployments with all combinations of those fields.   

Better, I think, to limit the options as far as possible while we get everything working for your use cases.

 * https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.15/#persistentvolumeclaimspec-v1-core

I also don't understand why Kubernetes authors chose to make `volumeClaimTemplate PersistentVolumeClaim` rather than introducing a new more limited template type. 

Having said all that, perhaps I need to write some prototype documentation for that snapshot restore case, so that we can understand how it'll work.

Reply:
I've added an EtcdPeerStorage type, as suggested, and we can discuss it today.